LEGO1/tgl/d3drm/texture.cpp

```diff
@@ -0,0 +1,202 @@
+#include "impl.h"
+
+using namespace TglImpl;
+
+DECOMP_SIZE_ASSERT(TglD3DRMIMAGE, 0x40);
+
+inline TglD3DRMIMAGE* TextureGetImage(IDirect3DRMTexture* p_texture)
+{
+	return reinterpret_cast<TglD3DRMIMAGE*>(p_texture->GetAppData());
+}
+
+// Forward declare to satisfy order check
+void TextureDestroyCallback(IDirect3DRMObject* pObject, void* pArg);
+
+// OFFSET: LEGO1 0x100a12a0
+Result TextureImpl::SetImage(IDirect3DRMTexture* p_self, TglD3DRMIMAGE* p_image)
+{
+	unsigned long appData;
+	Result result;
+
+	appData = reinterpret_cast<unsigned long>(p_image);
+
+	// This is here because in the original code they asserted
+	// on the return value being NULL.
+	TextureGetImage(p_self);
+
+	result = ResultVal(p_self->SetAppData(appData));
+	if (Succeeded(result) && p_image) {
+		result = ResultVal(p_self->AddDestroyCallback(TextureDestroyCallback, NULL));
+		if (!Succeeded(result)) {
+			p_self->SetAppData(0);
+		}
+	}
+	return result;
+}
+
+// OFFSET: LEGO1 0x100a1300
+void TextureDestroyCallback(IDirect3DRMObject* pObject, void* pArg)
+{
+	TglD3DRMIMAGE* pImage = reinterpret_cast<TglD3DRMIMAGE*>(pObject->GetAppData());
+	delete pImage;
+	pObject->SetAppData(0);
+}
+
+// OFFSET: LEGO1 0x100a1330
+TglD3DRMIMAGE::TglD3DRMIMAGE(
+	int p_width,
+	int p_height,
+	int p_depth,
+	void* p_buffer,
+	int p_useBuffer,
+	int p_paletteSize,
+	PaletteEntry* p_palette
+)
+{
+	m_image.aspectx = 1;
+	m_image.aspecty = 1;
+	m_image.width = 0;
+	m_image.height = 0;
+	m_image.depth = 0;
+	m_image.rgb = 0;
+	m_image.bytes_per_line = 0;
+	m_image.buffer1 = NULL;
+	m_image.buffer2 = NULL;
+	m_image.red_mask = 0xFF;
+	m_image.green_mask = 0xFF;
+	m_image.blue_mask = 0xFF;
+	m_image.alpha_mask = 0xFF;
+	m_image.palette_size = 0;
+	m_image.palette = NULL;
+	m_texelsAllocatedByClient = 0;
+	if (p_buffer != NULL) {
+		CreateBuffer(p_width, p_height, p_depth, p_buffer, p_useBuffer);
+	}
+	if (p_palette != NULL) {
+		InitializePalette(p_paletteSize, p_palette);
+	}
+}
+
+// OFFSET: LEGO1 0x100a13b0
+void TglD3DRMIMAGE::Destroy()
+{
+	if (m_texelsAllocatedByClient == 0) {
+		free(m_image.buffer1);
+	}
+	free(m_image.palette);
+}
+
+// OFFSET: LEGO1 0x100a13e0 STUB
+Result TglD3DRMIMAGE::CreateBuffer(int p_width, int p_height, int p_depth, void* p_buffer, int p_useBuffer)
+{
+	return Error;
+}
+
+// OFFSET: LEGO1 0x100a1510
+void TglD3DRMIMAGE::FillRowsOfTexture(int p_y, int p_height, char* p_content)
+{
+	// The purpose is clearly this but I can't get the assembly to line up.
+	memcpy((char*) m_image.buffer1 + (p_y * m_image.bytes_per_line), p_content, p_height * m_image.bytes_per_line);
+}
+
+// OFFSET: LEGO1 0x100a1550
+Result TglD3DRMIMAGE::InitializePalette(int p_paletteSize, PaletteEntry* p_palette)
+{
+	// This function is a 100% match if the PaletteEntry class is copied
+	// into into the TglD3DRMIMAGE class instead of being a global struct.
+	if (m_image.palette_size != p_paletteSize) {
+		if (m_image.palette != NULL) {
+			free(m_image.palette);
```

Comment:
This should be `delete`

Reply:
When substituing `free` with `delete`, the match moves from 68% to a full 100%:

![image](https://github.com/isledecomp/isle/assets/1135351/9ab53c22-6ed2-45b4-b499-4c84be8c7b30)

Again, I might be missing something here, but `delete` just seems correct.

Reply:
How did you make this match?? I changed the code to new/delete but it still does not for me.